File: app/src/main/java/me/ykrank/s1next/view/dialog/BlacklistDialogFragment.java
package me.ykrank.s1next.view.dialog;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.text.TextUtils;
import android.view.WindowManager;

import me.ykrank.s1next.R;
import me.ykrank.s1next.data.db.dbmodel.BlackList;
import me.ykrank.s1next.databinding.DialogBlacklistBinding;
import me.ykrank.s1next.view.internal.RequestCode;
import me.ykrank.s1next.viewmodel.BlackListViewModel;

/**
 * A dialog lets the user add or edit blacklist.
 */
public final class BlacklistDialogFragment extends BaseDialogFragment {

    public static final String BLACKLIST_TAG = "blacklist";
    public static final String TAG = BlacklistDialogFragment.class.getName();

    private BlackList mBlacklist;

    public static BlacklistDialogFragment newInstance(BlackList blackList) {
        BlacklistDialogFragment fragment = new BlacklistDialogFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(BLACKLIST_TAG, blackList);
        fragment.setArguments(bundle);

        return fragment;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Activity activity = getActivity();
        Object blacklist = getArguments().get(BLACKLIST_TAG);
        if (blacklist != null) {
            mBlacklist = (BlackList) blacklist;
        }

        DialogBlacklistBinding binding = DataBindingUtil.inflate(activity.getLayoutInflater(),
                R.layout.dialog_blacklist, null, false);
        BlackListViewModel blackListViewModel = new BlackListViewModel();
        if (mBlacklist != null)
            blackListViewModel.blacklist.set(mBlacklist);
        binding.setBlackListViewModel(blackListViewModel);

        AlertDialog alertDialog = new AlertDialog.Builder(activity)
                .setTitle(mBlacklist == null ? R.string.menu_blacklist_add : R.string.menu_blacklist_edit)
                .setView(binding.getRoot())
                .setPositiveButton(android.R.string.ok, (dialog, which) ->
                {
                    String authorIds = binding.blacklistId.getText().toString().trim();
                    int authorId = TextUtils.isEmpty(authorIds) ? 0 : Integer.parseInt(authorIds);
                    String authorName = binding.blacklistName.getText().toString();
                    @BlackList.ForumFLag int forum = binding.switchForum.isChecked() ? BlackList.HIDE_FORUM : BlackList.NORMAL;
                    @BlackList.PostFLag int post = binding.switchForum.isChecked() ? BlackList.HIDE_POST : BlackList.NORMAL;
                    BlackList blackList = new BlackList(authorId, authorName, post, forum);
                    Intent intent = new Intent();
                    intent.putExtra(BLACKLIST_TAG, blackList);
                    getTargetFragment().onActivityResult(RequestCode.REQUEST_CODE_BLACKLIST, Activity.RESULT_OK, intent);
                })
                .setNegativeButton(android.R.string.cancel, null)
                .create();
        alertDialog.getWindow().setSoftInputMode(
                WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE);
        return alertDialog;
    }
}

File: app/src/main/java/me/ykrank/s1next/view/dialog/BlacklistDialogFragment.kt
package me.ykrank.s1next.view.dialog

import android.app.Activity
import android.app.Dialog
import android.content.Intent
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import android.text.TextUtils
import android.view.WindowManager
import androidx.annotation.IdRes

import me.ykrank.s1next.R
import me.ykrank.s1next.data.db.dbmodel.BlackList
import me.ykrank.s1next.databinding.DialogBlacklistBinding
import me.ykrank.s1next.view.internal.RequestCode
import me.ykrank.s1next.viewmodel.BlackListViewModel

/**
 * A dialog lets the user add or edit blacklist.
 */
class BlacklistDialogFragment : BaseDialogFragment() {

    private var mBlacklist: BlackList? = null
    private var mBinding: DialogBlacklistBinding? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val activity = activity!!
        val blacklist = arguments?.get(BLACKLIST_TAG)
        if (blacklist != null) {
            mBlacklist = blacklist as BlackList
        }

        val binding = DataBindingUtil.inflate<DialogBlacklistBinding>(
            activity.layoutInflater,
            R.layout.dialog_blacklist, null, false
        )
        mBinding = binding
        val blackListViewModel = BlackListViewModel()
        if (mBlacklist != null)
            blackListViewModel.blacklist.set(mBlacklist)
        binding.blackListViewModel = blackListViewModel

        binding.radioGroupForum.check(forumRadioIdFromFlag())
        binding.radioGroupPost.check(postRadioIdFromFlag())

        val alertDialog = AlertDialog.Builder(activity)
            .setTitle(if (mBlacklist == null) R.string.menu_blacklist_add else R.string.menu_blacklist_edit)
            .setView(binding.root)
            .setPositiveButton(android.R.string.ok) { dialog, which ->
                val authorIds = binding.blacklistId.text.toString().trim { it <= ' ' }
                val authorId = if (TextUtils.isEmpty(authorIds)) 0 else Integer.parseInt(authorIds)
                val authorName = binding.blacklistName.text.toString()
                @BlackList.ForumFLag val forum = forumFlagFromRadioId()
                @BlackList.PostFLag val post = postFlagFromRadioId()
                val blackList = BlackList(authorId, authorName, post, forum)
                val intent = Intent()
                intent.putExtra(BLACKLIST_TAG, blackList)
                targetFragment?.onActivityResult(
                    RequestCode.REQUEST_CODE_BLACKLIST,
                    Activity.RESULT_OK,
                    intent
                )
            }
            .setNegativeButton(android.R.string.cancel, null)
            .create()
        alertDialog.window?.setSoftInputMode(
            WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE
        )
        return alertDialog
    }

    @IdRes
    private fun forumRadioIdFromFlag(): Int {
        return when (mBlacklist?.forum) {
            BlackList.HIDE_FORUM -> R.id.radio_forum_hide
            BlackList.DEL_FORUM -> R.id.radio_forum_del
            else -> R.id.radio_forum_normal
        }
    }

    @BlackList.ForumFLag
    private fun forumFlagFromRadioId(): Int {
        return when (mBinding?.radioGroupForum?.checkedRadioButtonId) {
            R.id.radio_forum_hide -> BlackList.HIDE_FORUM
            R.id.radio_forum_del -> BlackList.DEL_FORUM
            else -> BlackList.NORMAL
        }
    }

    @IdRes
    private fun postRadioIdFromFlag(): Int {
        return when (mBlacklist?.post) {
            BlackList.HIDE_POST -> R.id.radio_post_hide
            BlackList.DEL_POST -> R.id.radio_post_del
            else -> R.id.radio_forum_normal
        }
    }

    @BlackList.PostFLag
    private fun postFlagFromRadioId(): Int {
        return when (mBinding?.radioGroupPost?.checkedRadioButtonId) {
            R.id.radio_post_hide -> BlackList.HIDE_POST
            R.id.radio_post_del -> BlackList.DEL_POST
            else -> BlackList.NORMAL
        }
    }

    companion object {

        val BLACKLIST_TAG = "blacklist"
        val TAG = BlacklistDialogFragment::class.java.name

        fun newInstance(blackList: BlackList?): BlacklistDialogFragment {
            val fragment = BlacklistDialogFragment()
            val bundle = Bundle()
            bundle.putParcelable(BLACKLIST_TAG, blackList)
            fragment.arguments = bundle

            return fragment
        }
    }
}
